File: Feature_Extraction/CFP_filterbank.py
# -*- coding: utf-8 -*-
"""
Created on Tue Aug 22 01:08:46 2017

@author: lisu
"""

"""
Created on Thu Feb. 21 09:58:42 2017

@author: lisu + icwei
"""

# import all library here
#from scipy import signal
import soundfile as sf
#from scipy.fftpack import fft, fftshift
#from datetime import datetime
import numpy as np
import scipy
#import stft                   # import costmized stft function
#import med_filter2            # import costmized function med_filter2
import matplotlib.pyplot as plt

# [tfr, f, t, N] = STFT(x, fr, fs, Hop, h)
def STFT(x, fr, fs, Hop, h):        
    t = np.arange(Hop, np.ceil(len(x)/float(Hop))*Hop, Hop)
    N = int(fs/float(fr))
    window_size = len(h)
    f = fs*np.linspace(0, 0.5, np.round(N/2), endpoint=True)
    Lh = int(np.floor(float(window_size-1) / 2))
    tfr = np.zeros((int(N), len(t)), dtype=np.float)     
    if len(x.shape) > 1: x = x[:,0]
        
    for icol in range(0, len(t)):
        ti = int(t[icol])           
        tau = np.arange(int(-min([round(N/2.0)-1, Lh, ti-1])), \
                        int(min([round(N/2.0)-1, Lh, len(x)-ti])))
        indices = np.mod(N + tau, N) + 1
        tfr[indices-1, icol] = x[ti+tau-1] * h[Lh+tau-1]/np.linalg.norm(h[Lh+tau-1])         
                            
    tfr = abs(scipy.fftpack.fft(tfr, n=N, axis=0))  
    return tfr, f, t, N

def nonlinear_func(X, g, cutoff):
    cutoff = int(cutoff)
    if g!=0:
        X[X<0] = 0
        X[:cutoff, :] = 0
        X[-cutoff:, :] = 0
        X = np.power(X, g)
    else:
        X = np.log(X)
        X[:cutoff, :] = 0
        X[-cutoff:, :] = 0
    return X

def Freq2LogFreqMapping(tfr, f, fr, fc, tc, NumPerOct):
    StartFreq = fc
    StopFreq = 1/tc
    Nest = int(np.ceil(np.log2(StopFreq/StartFreq))*NumPerOct)
    central_freq = []

    for i in range(0, Nest):
        CenFreq = StartFreq*pow(2, float(i)/NumPerOct)
        if CenFreq < StopFreq:
            central_freq.append(CenFreq)
        else:
            break

    Nest = len(central_freq)
    freq_band_transformation = np.zeros((Nest-1, len(f)), dtype=np.float)
    for i in range(1, Nest-1):
        l = int(round(central_freq[i-1]/fr))
        r = int(round(central_freq[i+1]/fr)+1)
        #rounding1
        if l >= r-1:
            freq_band_transformation[i, l] = 1
        else:
            for j in range(l, r):
                if f[j] > central_freq[i-1] and f[j] < central_freq[i]:
                    freq_band_transformation[i, j] = (f[j] - central_freq[i-1]) / (central_freq[i] - central_freq[i-1])
                elif f[j] > central_freq[i] and f[j] < central_freq[i+1]:
                    freq_band_transformation[i, j] = (central_freq[i + 1] - f[j]) / (central_freq[i + 1] - central_freq[i])
    tfrL = np.dot(freq_band_transformation, tfr)
    return tfrL, central_freq

def Quef2LogFreqMapping(ceps, q, fs, fc, tc, NumPerOct):
    StartFreq = fc
    StopFreq = 1/tc
    Nest = int(np.ceil(np.log2(StopFreq/StartFreq))*NumPerOct)
    central_freq = []

    for i in range(0, Nest):
        CenFreq = StartFreq*pow(2, float(i)/NumPerOct)
        if CenFreq < StopFreq:
            central_freq.append(CenFreq)
        else:
            break
    f = 1/q
    Nest = len(central_freq)
    freq_band_transformation = np.zeros((Nest-1, len(f)), dtype=np.float)
    for i in range(1, Nest-1):
        for j in range(int(round(fs/central_freq[i+1])), int(round(fs/central_freq[i-1])+1)):
            if f[j] > central_freq[i-1] and f[j] < central_freq[i]:
                freq_band_transformation[i, j] = (f[j] - central_freq[i-1])/(central_freq[i] - central_freq[i-1])
            elif f[j] > central_freq[i] and f[j] < central_freq[i+1]:
                freq_band_transformation[i, j] = (central_freq[i + 1] - f[j]) / (central_freq[i + 1] - central_freq[i])

    tfrL = np.dot(freq_band_transformation, ceps)
    return tfrL, central_freq

def peakPicking(data):
    M, N = np.shape(data)
    pre = data[1:M - 1, :] - data[0:M - 2, :]
    pre[pre < 0] = 0
    pre[pre > 0] = 1

    #    sft_matrix_post = np.append(ceps[2:,:], np.zeros([1,N]), axis=0)
    post = data[1:M - 1, :] - data[2:, :]
    post[post < 0] = 0
    post[post > 0] = 1

    mask = pre * post
    ext_mask = np.append(np.zeros([1, N]), mask, axis=0)
    ext_mask = np.append(ext_mask, np.zeros([1, N]), axis=0)
    rdata = data * ext_mask
    return rdata
        
def CFP_filterbank(x, fr, fs, Hop, h, fc, tc, g, NumPerOctave):
    NumofLayer = np.size(g)

    [tfr, f, t, N] = STFT(x, fr, fs, Hop, h)
    tfr = np.power(abs(tfr), g[0])
    ceps = np.zeros(tfr.shape)


    if NumofLayer >= 2:
        for gc in range(1, NumofLayer):
            if np.remainder(gc, 2) == 1:
                tc_idx = round(fs*tc)
                ceps = np.real(np.fft.fft(tfr, axis=0))/np.sqrt(N)
                ceps = nonlinear_func(ceps, g[gc], tc_idx)
            else:
                fc_idx = round(fc/fr)
                tfr = np.real(np.fft.fft(ceps, axis=0))/np.sqrt(N)
                tfr = nonlinear_func(tfr, g[gc], fc_idx)

    tfr = tfr[:int(round(N/2)),:]
    ceps = ceps[:int(round(N/2)),:]

    HighFreqIdx = int(round((1/tc)/fr)+1)
    f = f[:HighFreqIdx]
    tfr = tfr[:HighFreqIdx,:]
    HighQuefIdx = int(round(fs/fc)+1)
    q = np.arange(HighQuefIdx)/float(fs)
    ceps = ceps[:HighQuefIdx,:]
    

    tfrLF, central_frequencies = Freq2LogFreqMapping(tfr, f, fr, fc, tc, NumPerOctave)
    tfrLQ, central_frequencies = Quef2LogFreqMapping(ceps, q, fs, fc, tc, NumPerOctave)

    return tfrLF, tfrLQ, f, q, t, central_frequencies 


# run "main" program here
if __name__== "__main__":
    
    x, fs = sf.read('opera_fem2.wav')
    x = scipy.signal.resample_poly(x, 16000, 44100)
    fs = 16000.0
    x = x.astype('float32')
    Hop = 320 
    h = scipy.signal.blackmanharris(2049)
    fr = 2.0
    frame_number = 170 
    fc = 80.0 # the frequency of the lowest pitch
    tc = 1/800.0 # the period of the highest pitch
    g = np.array([0.24, 0.6, 1])
    NumPerOctave = 48
    
    tfrLF, tfrLQ, f, q, t, central_frequencies = CFP_filterbank(x, fr, fs, Hop, h, fc, tc, g, NumPerOctave)
    Z = tfrLF*tfrLQ
#    np.save(filelist[i], Z)
    # show piano roll
#    matplotlib qt
    plt.figure(1)
    plt.imshow(tfrLF)
    plt.figure(2)
    plt.imshow(tfrLQ)
    
    
    
    

File: OnGithub/Feature_Extraction/CFP_filterbank.m
% output modes:
% 1) time-frequency (f-based and q-based)
% 2) time-log-frequency, with overlapped filterbank (f-based and q-based) 
% 3) semotone, max pooling and without filterbank (f-based and q-based)
function [tfrLF, tfrLQ, t, central_frequencies] = CFP_filterbank(x, fr, fs, Hop, h, fc, tc, g, NumPerOctave)

NumofLayer = length(g);

[tfr, f, t, N] = STFT(x, fr, fs, Hop, h);

tfr = abs(tfr.^g(1));

ceps = 0;
if NumofLayer >= 2
    for gc = 2:NumofLayer
        if rem(gc, 2)==0
            tc_idx = round(fs*tc);
            ceps = real(fft(tfr))./sqrt(N);
            ceps = nonlinear_func(ceps, g(gc), tc_idx);
        else
            fc_idx = round(fr/fc);
            tfr = real(fft(ceps))./sqrt(N);
            tfr = nonlinear_func(tfr, g(gc), fc_idx);
        end    
    end
end

tfr = tfr(1:round(N/2),:);
HighFreqIdx = round((1/tc)/fr)+1;
f = f(1:HighFreqIdx);
tfr = tfr(1:HighFreqIdx,:);
[tfrLF, central_frequencies.LF] = FreqToLogFreqMapping(tfr, f, fr, fc, tc, NumPerOctave);

tfrLQ = ceps;
if NumofLayer >= 2
    ceps = ceps(1:round(N/2),:);
    HighQuefIdx = round(fs/fc)+1;
    q = (0:HighQuefIdx-1)./fs;
    ceps = ceps(1:HighQuefIdx,:);
    [tfrLQ, central_frequencies.LQ] = QuefToLogFreqMapping(ceps, q, fs, fc, tc, NumPerOctave);
end
end

function [tfr, f, t, N] = STFT(x, fr, fs, Hop, h)
% STFT by Li Su, 2017
% Reference:
% Time-frequency toolbox by Patrick Flanderin
% fr: desired frequency resolution
% fs: sampling frequency
% Note: fr = alpha * fs ->
% alpha = fr / fs
% tfr: optput STFT (full frequency)
% f: output frequency grid (only positive frequency) 
% t: output time grid (from the start to the end)

if size(h,2) > size(h,1)
    h = h';
end

	% for tfr
alpha = fr/fs;
N = length(-0.5+alpha:alpha:0.5);
Win_length = max(size(h));
f = fs.*linspace(0, 0.5, round(N/2))' ;

Lh = floor((Win_length-1)/2);
t = Hop:Hop:floor(length(x)/Hop)*Hop;
x_Frame = zeros(N, length(t));
for ii = 1:length(t)
    ti = t(ii); 
    tau = -min([round(N/2)-1,Lh,ti-1]):min([round(N/2)-1,Lh,length(x)-ti]);
    indices= rem(N+tau,N)+1;
    norm_h=norm(h(Lh+1+tau)); 
    
    A = x(ti+tau)-mean(x(ti+tau));
    B = conj( h(Lh+1+tau));
    
    if size(A, 1) ~= size(B, 1)
        x_Frame(indices,ii) = A .* B' / norm_h;
    else
        x_Frame(indices,ii) = A .* B / norm_h;
    end
end

tfr = fft(x_Frame, N, 1);
end

function X = nonlinear_func(X, g, cutoff)
if g~=0
    X(X<0) = 0;
    X(1:cutoff, :) = 0;
    X(end-cutoff+1:end, :) = 0;
    X = X.^g;
else
    X = log(X);
    X(1:cutoff, :) = 0;
    X(end-cutoff+1:end, :) = 0;
end
end

function [tfrL, central_frequencies] = FreqToLogFreqMapping(tfr, f, fr, fc, tc, NumPerOctave)
StartFreq = fc; StopFreq = 1/tc;
Nest = ceil(log2(StopFreq/StartFreq))*NumPerOctave;

central_frequencies = [];
for i = 1:Nest
    CenFreq = StartFreq*2^((i-1)/NumPerOctave);
    if CenFreq < StopFreq
        central_frequencies = [central_frequencies; CenFreq];
    else
        break;
    end
end

Nest = length(central_frequencies);
freq_band_transformation = zeros(Nest-1, length(f));
for i = 2:length(central_frequencies)-1
    for j = round(central_frequencies(i-1)/fr+1) : round(central_frequencies(i+1)/fr+1)
        if (f(j)>central_frequencies(i-1) && f(j)<central_frequencies(i))
            freq_band_transformation(i,j) = (f(j)-central_frequencies(i-1))/(central_frequencies(i)-central_frequencies(i-1));%/D;
        elseif (f(j)>central_frequencies(i) && f(j)<central_frequencies(i+1))
            freq_band_transformation(i,j) = (central_frequencies(i+1)-f(j))/(central_frequencies(i+1)-central_frequencies(i));%/D;
        end
    end
end

tfrL = freq_band_transformation*tfr;
end

function [tfrL, central_frequencies] = QuefToLogFreqMapping(ceps, q, fs, fc, tc, NumPerOctave)
StartFreq = fc; StopFreq = 1/tc;
Nest = ceil(log2(StopFreq/StartFreq))*NumPerOctave;

central_frequencies = [];
for i = 1:Nest
    CenFreq = StartFreq*2^((i-1)/NumPerOctave);
    if CenFreq < StopFreq
        central_frequencies = [central_frequencies; CenFreq];
    else
        break;
    end
end

f = 1./q;
Nest = length(central_frequencies);
freq_band_transformation = zeros(Nest-1, length(q));
for i = 2:length(central_frequencies)-1
    for j = round(fs/central_frequencies(i-1)+1) :-1: round(fs/central_frequencies(i+1)+1)
        if (f(j)>central_frequencies(i-1) && f(j)<central_frequencies(i))
            freq_band_transformation(i,j) = (f(j)-central_frequencies(i-1))/(central_frequencies(i)-central_frequencies(i-1));%/D;
        elseif (f(j)>central_frequencies(i) && f(j)<central_frequencies(i+1))
            freq_band_transformation(i,j) = (central_frequencies(i+1)-f(j))/(central_frequencies(i+1)-central_frequencies(i));%/D;
        end
    end
end

tfrL = freq_band_transformation*ceps;
end